### DISCUSSION TITLE
✈ Player Warps

### DISCUSSION THREAD

**Discussion by ghost** [Zrips/CMI#6049] [Feature Requests]
CMI is almost a paragon of teleportation. It has a plethora of "/tp" commands, it has its own portal system, it has its own home system, it has its own Admin warps system.
Why not add Player Warps?

Trust me, this feature would be the best thing ever! CMI's Player Warp system can be inspired by plugins like "[HyperDrive](https://www.spigotmc.org/resources/%E2%96%BAhyperdrive-advanced-teleportation-plugin-%E2%96%BA.17184/)" and "[Warp & Teleport System](https://www.spigotmc.org/resources/premium-warps-portals-and-more-warp-teleport-system-1-8-1-17.66035/)".

"Why not use those plugins instead?"
Well there can be any number of reasons for any number of people. Both are great plugins, but there can be a lot of overlap in features and a lack of integration with CMIs teleportation.
Sometimes you got to draw the line when it comes to how many plugins are installed with similar functionality.

### How it Works:
You can have a config file specifically for Player warps. If not, you can simply place it in "config.yml" with the normal warp feature.
Have an option to enable/disable player warps.

Player warps can act as an alternate Home system, but one does not NEED to replace the other. Like homes but with options and functionality.
Player warps can be a way for players to connect with each other in a way that is not seen on most servers.
Of course there has to be a GUI, where all of the warps are. Players can filter between the admin warps and player warps. Even search for warps by name or by owner. Even searching by popularity.
Think of the vast possibilities.

Player warps can have various settings such as:
- Being private warps, only accessible by the owner. Public warps, accessible to all players. Blacklist/Whitelist.
- Setting a bed warp for respawning.
- Setting a name and a description for the warp.
- Setting a price for other players to use the warp (on top of a base price set to use warps). The extra price for players to use the warp can go to the owner of the warp.
- A rating system for players to rate other players' warps, such as likes/dislikes, popularity meter, ETC.
- Renaming warps, changing warp descriptions, changing location, ETC.
- Of course, there can be configured economy/EXP/item costs for creating warps and setting/changing their options.
- Players can pay to have their warp featured in a filter.

### Cross Server Warps:
Can't forget cross server warps. You can add an option in the warp section config to enable cross server warping for admin warps and one for player warps.
Perhaps warps can be saved in the MySQL database, and CMIs velocity plugin can support it too.
Some servers are designed to support or require cross server warp functionality. No reason not to have it.

The possibilities, utilities, versatilities, functionalities, and usability of player warps are endless.
Player warps are the way of the future, and CMI is already half way there.

**Comment by ghost** [comment]
uh... anyone?

**Comment by mrfloris** [comment]
<img width="967" alt="Screenshot 2021-09-08 at 09 06 26" src="https://user-images.githubusercontent.com/28841349/132462487-0fc6bb7c-de4a-471c-8888-9322476aaa2c.png">

While CMI already covers player-warps through the permission system, it's not 100% how you suggest it probably. 

I do like the idea that the CMI Home system could be extended to let someone edit their home(s) to be a visit, unless set private? 
I also like the idea that player warps automatically go to a certain group, and are perhaps stored separately, but I don't think it's a big deal if all warps are in a single warps.yml file.
